chore(deps): Bump `@metamask/phishing-controller` from `^9.0.0` to `^12.0.1`

MajorLift opened the pull request:
## **Description**

```md
### Changed

- **BREAKING:** Bump `@metamask/phishing-controller` from `^9.0.0` to `^12.0.1` ([#11111](https://github.com/MetaMask/metamask-mobile/pull/11111))
```

## **Related issues**

- Closes #11046

## **Manual testing steps**

## **Screenshots/Recordings**

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

MajorLift committed:
Add new `PhishingController` property to `initial-background-state` object

MajorLift committed:
Revert "Restore `@ts-expect-error` directive for `PhishingControllerMessenger`"

This reverts commit 47d002b5931c84d0435ae996a851a2813f4caec6.

MajorLift committed:
Bump `@metamask/phishing-controller` from `^9.0.0` to `^12.0.1`

mikesposito commented on yarn.lock:
```diff
@@ -4846,7 +4846,21 @@
+    fastest-levenshtein "^1.0.16"
+    punycode "^2.1.1"
+
+"@metamask/phishing-controller@^9.0.1":
```
Hmm, it looks like we still have ^9.0.1 in the tree

MajorLift replied:
Good catch! Looks like we have v9 and v10 via:
- `eth-snap-keyring@^4.0.0` > `snaps-controllers@^8.1.1` > `phishing-controller@^9.0.1`
- `snaps-controllers@^9.4.0` > `phishing-controller@^10.1.1`

I'm not sure how far we should go with aligning transitive dependencies in the context of the controllers upgrade OKR. Probably something to discuss in the next retro or OKR workshop.

I'll try updating `eth-snap-keyring`, `snaps-controllers` (and peer dep `snaps-execution-environments`), and see how far I get.

mikesposito replied:
[These types](https://github.com/MetaMask/snaps/blob/de6fbb16c55feaa13190aeec0b017b4818bf7617/packages/snaps-controllers/src/interface/SnapInterfaceController.ts#L73) imported in `SnapInterfaceController` are `PhishingController`'s messenger actions, so I guess that the action handler will be the one instantiated on Mobile.

However, by looking at the PhishingController changelog I don't see anything alarming

MajorLift replied:
`SnapInterfaceController` is expecting the actions as they are defined in `phishing-controller@9.0.0`, so if the action names, action handlers, or methods have changed between v9 and v12 that might cause issues. I'm taking a look at the following:

- https://github.com/MetaMask/core/compare/%40metamask/phishing-controller%409.0.0...%40metamask/phishing-controller%4012.0.2
  - https://github.com/MetaMask/core/blob/02b5a7ae2208a397e97c9f78266ec2bdb2259fcf/packages/phishing-controller/src/PhishingController.ts

It doesn't look like there's anything breaking at first glance, but it's not obvious since there have been a lot of changes in the method implementations. Good call on pointing this out. Definitely will be taking a closer look.

mikesposito replied:
PhishingController has been updated in snaps packages as well

mikesposito commented on yarn.lock:
```diff
@@ -4846,7 +4846,21 @@
+    "@metamask/controller-utils" "^11.1.0"
+    "@noble/hashes" "^1.4.0"
+    "@types/punycode" "^2.1.0"
+    eth-phishing-detect "^1.2.0"
```
Uh, interesting, this should not be here anymore. Not related to your PR, but I just opened another PR upstream to remove it: https://github.com/MetaMask/core/pull/4681

MajorLift committed:
Bump `@metamask/base-controller` from `^6.0.0` to `^7.0.0`

MajorLift committed:
Add temporary `ts-expect-error` directives for base-controller version mismatches

- These are because mobile's base-controller version is now ahead of the version used in the controllers
- These will be removed during the controller upgrades

MajorLift committed:
Replace all `ts-ignore` base-controller version mismatch directives with `ts-expect-error`

- There is never a reason to use `ts-ignore` over `ts-expect-error`, as it fails silently when the error it is suppressing is removed.

MajorLift committed:
Add `ts-expect-error` directive for `SelectedNetworkController`

MajorLift committed:
Add `PhishingControllerEvents` to `GlobalEvents` and `PhishingController:stateChange` to `datamodel` allowlist

mikesposito committed:
Merge branch 'main' into jongsun/chore/bump/phishing-controller-12.0.1

mikesposito committed:
chore: use `@metamask/phishing-controller@^12.0.3`

mikesposito committed:
chore: deduplicate packages

mikesposito committed:
fix: remove unused `ts-expect-error` directive

mikesposito committed:
Merge branch 'main' into jongsun/chore/bump/phishing-controller-12.0.1

mikesposito committed:
Merge branch 'main' into jongsun/chore/bump/phishing-controller-12.0.1

cryptodev-2s committed:
Merge branch 'main' into jongsun/chore/bump/phishing-controller-12.0.1

mikesposito committed:
Merge branch 'main' into jongsun/chore/bump/phishing-controller-12.0.1

mikesposito committed:
Merge branch 'main' into jongsun/chore/bump/phishing-controller-12.0.1

mikesposito committed:
Merge branch 'main' into jongsun/chore/bump/phishing-controller-12.0.1

mikesposito committed:
Merge branch 'main' into jongsun/chore/bump/phishing-controller-12.0.1

mikesposito committed:
Merge branch 'main' into jongsun/chore/bump/phishing-controller-12.0.1

mikesposito committed:
fix: use origin instead of hostname

mikesposito committed:
test: change mock to accomodate origin

mikesposito committed:
refactor: rename `isAllowedUrl` in `isAllowedOrigin`

mikesposito committed:
test: cover url with missing proto

mikesposito commented:
> Failed conditions
 [25.9% Coverage on New Code](https://sonarcloud.io/component_measures?id=metamask-mobile&pullRequest=11111&metric=new_coverage&view=list) (required ≥ 40%)
 
 I'm a bit confused by [the lines](https://sonarcloud.io/component_measures?id=metamask-mobile&pullRequest=11111&metric=new_coverage&view=list) showed as "added" in Sonar, but some methods that were originally uncovered have been renamed.

mikesposito committed:
Merge branch 'main' into jongsun/chore/bump/phishing-controller-12.0.1

mikesposito committed:
refactor: remove unused ts directives

mikesposito commented on app/core/Engine.ts:
```diff
@@ -903,20 +893,14 @@ class Engine {
         this.controllerMessenger,
         'SnapController:updateSnapState',
       ),
-      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-      // @ts-ignore
```
There are some other directives with no description a few lines above these, but I'm leaving them there since they are unrelated to PhishingController

Cal-L replied:
bypassed

cryptodev-2s commented on app/util/linkCheck.ts:
```diff
@@ -8,12 +8,12 @@ const DENYLISTED_DOMAINS = ['metamask.app.link'];
   ALLOWED_PROTOCOLS.includes(protocol);
 
-const isAllowedHostname = (hostname: string): boolean => {
+const isAllowedUrl = ({ hostname, origin }: Url<string>): boolean => {
```
I noticed that the hostname isn't being used here. We can simplify the function by having it take an origin directly instead of an Object URL. Also, we can rename it to `isAllowedOrigin`.

mikesposito replied:
hostname in this function is [used here](https://github.com/MetaMask/metamask-mobile/pull/11111/files#diff-96acc327df0b686bdfb0adf7e79c36ef0879bd95e02822c8da215e7ac36c5980R19), since `DENYLISTED_DOMAINS` includes hostnames

mikesposito committed:
Merge branch 'main' into jongsun/chore/bump/phishing-controller-12.0.1

mikesposito committed:
Merge branch 'main' into jongsun/chore/bump/phishing-controller-12.0.1

Cal-L reviewed (approved):
LGTM

cryptodev-2s committed:
Merge branch 'main' into jongsun/chore/bump/phishing-controller-12.0.1
